apps/emqx_rule_engine/src/emqx_rule_actions.erl

```diff
@@ -188,7 +201,7 @@ on_action_republish(Selected, _Envs = #{
             id = emqx_guid:gen(),
             qos = if TargetQoS =:= -1 -> 0; true -> TargetQoS end,
             from = ActId,
-            flags = #{dup => false, retain => false},
+            flags = #{dup => false, retain => if TargetRetain =:= 0 -> false; true -> true end},
             headers = #{republish_by => ActId},
             topic = emqx_rule_utils:proc_tmpl(TopicTks, Selected),
             payload = emqx_rule_utils:proc_tmpl(PayloadTks, Selected),
```

Comment:
Instead of closing this PR and opening another one, you could push your changes to the same branch as this PR used, so the comments above won't lost.